[WPE][GTK] Drop libnotify

carlosgcampos opened the pull request:
#### 95858e6cc2e3f804cc95b4f55cd7532329db790a
<pre>
[WPE][GTK] Drop libnotify
<a href="https://bugs.webkit.org/show_bug.cgi?id=240347">https://bugs.webkit.org/show_bug.cgi?id=240347</a>

Reviewed by Michael Catanzaro.

Add NotificationService singleton to communicate with the notifications
service.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:
(webkitNotificationCreate):
(webkitNotificationGetWebNotification):
(webkit_notification_get_id):
(webkit_notification_get_title):
(webkit_notification_get_body):
(webkit_notification_get_tag):
(webkitNotificationGetWebView): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitNotificationPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
(WebKitNotificationProvider::~WebKitNotificationProvider):
(WebKitNotificationProvider::apiNotificationCloseCallback):
(WebKitNotificationProvider::apiNotificationClickedCallback):
(WebKitNotificationProvider::closeAPINotification):
(WebKitNotificationProvider::withdrawAnyPreviousAPINotificationMatchingTag):
(WebKitNotificationProvider::show):
(WebKitNotificationProvider::cancelNotificationByID):
(WebKitNotificationProvider::didClickNotification):
(WebKitNotificationProvider::didCloseNotification):
(WebKitNotificationProvider::notificationCloseCallback): Deleted.
(WebKitNotificationProvider::notificationClickedCallback): Deleted.
(WebKitNotificationProvider::withdrawAnyPreviousNotificationMatchingTag): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(notifyNotificationClosed): Deleted.
(notifyNotificationClicked): Deleted.
(webNotificationClosed): Deleted.
(webkitWebViewShowNotification): Deleted.
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp: Added.
(WebKit::NotificationService::singleton):
(WebKit::NotificationService::NotificationService):
(WebKit::NotificationService::processCapabilities):
(WebKit::NotificationService::showNotification):
(WebKit::NotificationService::cancelNotification):
(WebKit::NotificationService::setNotificationID):
(WebKit::NotificationService::findNotification):
(WebKit::NotificationService::handleSignal):
(WebKit::NotificationService::didClickNotification):
(WebKit::NotificationService::didCloseNotification):
(WebKit::NotificationService::addObserver):
(WebKit::NotificationService::removeObserver):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.h: Added.
* Source/cmake/FindLibNotify.cmake: Removed.
* Source/cmake/OptionsGTK.cmake:

Canonical link: <a href="https://commits.webkit.org/253179@main">https://commits.webkit.org/253179@main</a>
</pre>

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:
```diff
@@ -51,12 +51,11 @@ enum {
 };
 
 struct _WebKitNotificationPrivate {
+    RefPtr<WebKit::WebNotification> notification;
```
Surely this should be:

```suggestion
    Ref<WebKit::WebNotification> notification;
```

because you're treating it as if it is not nullable everywhere.

carlosgcampos replied:
That requires a value provided in the constructor and that's not possible with GObject priv struct.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
```diff
@@ -80,69 +81,95 @@ WebKitNotificationProvider::WebKitNotificationProvider(WebNotificationManagerPro
         }
     }
 
 #ifndef NDEBUG
```
Good opportunity for a drive-by fix:

```suggestion
#if ASSERT_ENABLED
```

(`NDEBUG` controls glibc assertions, not WebKit assertions.)

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
```diff
@@ -80,69 +81,95 @@ WebKitNotificationProvider::WebKitNotificationProvider(WebNotificationManagerPro
-        // FIXME: glib needs to find their own solution to handling pageless notifications.
-        // I won't make something up for them.
+        // FIXME: glib API needs to find their own solution to handling pageless notifications.
+        show(webNotification);
```
So now we show the notification unconditionally, without allowing the application to control it? That's probably worse than dropping the notification, because it will bypass the application's permission store. Maybe the original behavior was better here?

carlosgcampos replied:
I don't think applications handle the permissions in show-notification callback, permissions are set in the context and handled by WebKit, and that's still working in this case, we don't even get here if permission hasn't been granted for the notification origin. We don't even expose the notification origin in the API.

mcatanzaro replied:
OK.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+
+    std::call_once(onceFlag, [] {
+        service.construct();
+    });
```
This class is not threadsafe, so you should just store a bool rather than using `std::call_once`. The point of `std::call_once` is to use a mutex to guarantee it gets initialized once regardless of thread.

carlosgcampos replied:
This is a common pattern in all singleton using LazyNeverDestroyed, I think it's harmless even if the class it not thread safe.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+    const char* objectPath = shouldUsePortal() ? "/org/freedesktop/portal/desktop" : "/org/freedesktop/Notifications";
+    const char* interfaceName = shouldUsePortal() ? "org.freedesktop.portal.Notification" : "org.freedesktop.Notifications";
+    GUniqueOutPtr<GError> error;
+    m_proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SESSION, G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES, nullptr, busName, objectPath, interfaceName, nullptr, &error.outPtr());
```
Why do you have to use sync I/O here? Wouldn't this be quite easy to convert to async?

TingPing replied:
Sync bus lookup is non-blocking after the first call. But the call below is blocking.

mcatanzaro replied:
What's the "first call?"

carlosgcampos replied:
Using async IO for these would make everything a lot more complicated for very little benefit. For the first notification show will happen before the proxy is constructed or capabilities processed, so we would need to handle the case and wait for the async to complete to show the notification. Since this only happens once I don't think it's worth it.

carlosgcampos replied:
> What's the "first call?"

I think he means getting the bus connection, because g_bus_get_sync returns a singleton, so only the first time it's called it needs to connect which is async. At this point, glib has already connected to the session bus for sure.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    if (!shouldUsePortal()) {
+        GRefPtr<GVariant> capabilities = adoptGRef(g_dbus_proxy_call_sync(m_proxy.get(), "GetCapabilities", nullptr, G_DBUS_CALL_FLAGS_NONE, s_dbusCallTimeout.millisecondsAs<int>(), nullptr, &error.outPtr()));
```
Ditto?

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+    if (!m_proxy)
+        return false;
+
+    auto findNotificationByTag = [&](const String& tag) -> Notification {
```
Better to capture only what you need, rather than everything. I think the only thing you need here is `this`, right?

carlosgcampos replied:
It's a lambda local to the function, but ok.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+        g_variant_builder_add(&builder, "{sv}", "default-action", g_variant_new_string("default"));
+        addResult.iterator->value.portalID = createVersion4UUIDString();
+        g_dbus_proxy_call(m_proxy.get(), "AddNotification", g_variant_new("(s@a{sv})", addResult.iterator->value.portalID.utf8().data(), g_variant_builder_end(&builder)),
+            G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```
Should pass a callback just to print a warning if it fails. Yes, it's slightly more effort, but the future benefit in case things break in the future will be worth it.

carlosgcampos replied:
Ok

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+
+        if (m_capabilities.contains(Capabilities::Actions)) {
+            g_variant_builder_add(&actionsBuilder, "s", "default");
+            g_variant_builder_add(&actionsBuilder, "s", _("Acknowledge"));
```
```suggestion
            // TRANSLATORS: the default action for a desktop notification created by a website.
            g_variant_builder_add(&actionsBuilder, "s", _("Acknowledge"));
```

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+        if (applicationID)
+            g_variant_builder_add(&hintsBuilder, "{sv}", "desktop-entry", g_variant_new_string(applicationID));
+
+        auto* value = static_cast<GValue*>(fastZeroedMalloc(sizeof(GValue)));
```
Can you use `UniqueRef` for this?

If not, you can use a `std::unique_ptr` with a custom deleter and `WTFMove` it into the callback.

carlosgcampos replied:
I'll try

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+        if (it->value.portalID.isEmpty())
+            return;
+
+        g_dbus_proxy_call(m_proxy.get(), "RemoveNotification", g_variant_new("(s)", it->value.portalID.utf8().data()), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```
I'd like to see callbacks everywhere you omit them currently. Even doing nothing except print error and return is valuable.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -0,0 +1,310 @@
+        if (!it->value.id)
+            return;
+
+        g_dbus_proxy_call(m_proxy.get(), "CloseNotification", g_variant_new("(u)", it->value.id), G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```
Ditto.

mcatanzaro reviewed (approved):
Nice, thanks.

carlosgcampos committed:
[WPE][GTK] Drop libnotify
https://bugs.webkit.org/show_bug.cgi?id=240347

Reviewed by Michael Catanzaro.

Add NotificationService singleton to communicate with the notifications
service.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:
(webkitNotificationCreate):
(webkitNotificationGetWebNotification):
(webkit_notification_get_id):
(webkit_notification_get_title):
(webkit_notification_get_body):
(webkit_notification_get_tag):
(webkitNotificationGetWebView): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitNotificationPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
(WebKitNotificationProvider::~WebKitNotificationProvider):
(WebKitNotificationProvider::apiNotificationCloseCallback):
(WebKitNotificationProvider::apiNotificationClickedCallback):
(WebKitNotificationProvider::closeAPINotification):
(WebKitNotificationProvider::withdrawAnyPreviousAPINotificationMatchingTag):
(WebKitNotificationProvider::show):
(WebKitNotificationProvider::cancelNotificationByID):
(WebKitNotificationProvider::didClickNotification):
(WebKitNotificationProvider::didCloseNotification):
(WebKitNotificationProvider::notificationCloseCallback): Deleted.
(WebKitNotificationProvider::notificationClickedCallback): Deleted.
(WebKitNotificationProvider::withdrawAnyPreviousNotificationMatchingTag): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(notifyNotificationClosed): Deleted.
(notifyNotificationClicked): Deleted.
(webNotificationClosed): Deleted.
(webkitWebViewShowNotification): Deleted.
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp: Added.
(WebKit::NotificationService::singleton):
(WebKit::NotificationService::NotificationService):
(WebKit::NotificationService::processCapabilities):
(WebKit::NotificationService::showNotification):
(WebKit::NotificationService::cancelNotification):
(WebKit::NotificationService::setNotificationID):
(WebKit::NotificationService::findNotification):
(WebKit::NotificationService::handleSignal):
(WebKit::NotificationService::didClickNotification):
(WebKit::NotificationService::didCloseNotification):
(WebKit::NotificationService::addObserver):
(WebKit::NotificationService::removeObserver):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.h: Added.
* Source/cmake/FindLibNotify.cmake: Removed.
* Source/cmake/OptionsGTK.cmake:

Canonical link: https://commits.webkit.org/253179@main
